kzg/src/msm/msm_impls.rs

```diff
@@ -0,0 +1,70 @@
+use crate::{G1Affine, G1Fp, G1GetFp, G1ProjAddAffine, Scalar256, G1};
+use alloc::vec::Vec;
+
+#[cfg(feature = "arkmsm")]
+use super::arkmsm::arkmsm_msm::VariableBaseMSM;
+
+#[cfg(not(feature = "arkmsm"))]
+use super::tiling_pippenger_ops::tiling_pippenger;
+
+#[cfg(feature = "parallel")]
+use super::tiling_parallel_pippenger::{parallel_affine_conv, tiling_parallel_pippenger};
+
+#[cfg(feature = "parallel")]
+pub fn msm_parallel<
+    TG1: G1 + G1GetFp<TG1Fp>,
+    TG1Fp: G1Fp,
+    TG1Affine: G1Affine<TG1, TG1Fp>,
+    TProjAddAffine: G1ProjAddAffine<TG1, TG1Fp, TG1Affine>,
+>(
+    points: &[TG1Affine],
+    scalars: &[Scalar256],
+) -> TG1 {
+    tiling_parallel_pippenger(points, scalars)
+}
+
+pub fn msm_sequential<
+    TG1: G1 + G1GetFp<TG1Fp>,
+    TG1Fp: G1Fp,
+    TG1Affine: G1Affine<TG1, TG1Fp>,
+    TProjAddAffine: G1ProjAddAffine<TG1, TG1Fp, TG1Affine>,
+>(
+    points: &[TG1Affine],
+    scalars: &[Scalar256],
+) -> TG1 {
+    #[cfg(not(feature = "arkmsm"))]
+    {
+        tiling_pippenger(points, scalars)
+    }
+
+    #[cfg(feature = "arkmsm")]
+    {
+        VariableBaseMSM::multi_scalar_mul::<TG1, TG1Fp, TG1Affine, TProjAddAffine>(points, scalars)
+    }
+}
+
+pub fn msm<
+    TG1: G1 + G1GetFp<TG1Fp>,
+    TG1Fp: G1Fp,
+    TG1Affine: G1Affine<TG1, TG1Fp>,
+    TProjAddAffine: G1ProjAddAffine<TG1, TG1Fp, TG1Affine>,
+>(
+    points: &[TG1Affine],
+    scalars: &[Scalar256],
+) -> TG1 {
+    #[cfg(feature = "parallel")]
+    return msm_parallel::<TG1, TG1Fp, TG1Affine, TProjAddAffine>(points, scalars);
+
+    #[cfg(not(feature = "parallel"))]
+    return msm_sequential::<TG1, TG1Fp, TG1Affine, TProjAddAffine>(points, scalars);
```

Comment:
msm() function should be called with affine + scalars. I think part of why this naive multiplication is done at g1_lincomb level is to bypass the affine & fr conversion cost. Not sure whether we gain anything by adding it here since callers will have already incurred the cost of conversion.

Reply:
Why not pass `G1` and `FsFr` into this function, and do conversions inside the `msm` function, if needed?

Reply:
makes sense, added public msm function which accepts G1 & Fr converts internally and calls affine version